controllers/serviceclaim_controller.go

```diff
@@ -47,13 +61,229 @@ type ServiceClaimReconciler struct {
+		}
+	}
+
+	// if the number of SED keys is more than the number of secret data entries
+	// that indicates one or more keys are missing
+	if len(sclaim.Spec.ServiceEndpointDefinitionKeys) > count {
+		sclaim.Status.State = "Pending"
+		if err := r.Status().Update(ctx, &sclaim); err != nil {
+			l.Error(err, "unable to update the ServiceClaim", "ServiceClaim", sclaim)
+			return "", err
+		}
+
+		return "", fmt.Errorf("Key not available in the list of SEDs")
+	}
+
+	// ServiceClassIdentity values are going to override
+	// any values in the secret resource
+	for _, sci := range sclaim.Spec.ServiceClassIdentity {
+		secret.StringData[sci.Name] = sci.Value
+	}
+
+	var cel primazaiov1alpha1.ClusterEnvironmentList
+	if err := r.List(ctx, &cel); err != nil {
+		l.Info("error fetching ClusterEnvironmentList", "error", err)
+		return "", client.IgnoreNotFound(err)
+	}
+
+	// loop over all ClusterEnvironment
+	for _, ce := range cel.Items {
+		// check if the ServiceClaim EnvironmentTag matches the EnvironmentName part of ClusterEnvironment
+		if sclaim.Spec.EnvironmentTag == ce.Spec.EnvironmentName {
+			sn := ce.Spec.ClusterContextSecret
+			k := client.ObjectKey{Namespace: ce.Namespace, Name: sn}
+			var s corev1.Secret
+			if err := r.Get(ctx, k, &s); err != nil {
+				return "", err
+			}
+
+			kubeconfig := s.Data["kubeconfig"]
+			cg, err := clientcmd.RESTConfigFromKubeConfig(kubeconfig)
+			if err != nil {
+				return "", err
+			}
+
+			cs, err := kubernetes.NewForConfig(cg)
+			if err != nil {
+				return "", err
+			}
+
+			_, err = cs.CoreV1().Secrets(ce.Namespace).Create(ctx, secret, metav1.CreateOptions{})
+			if err != nil {
+				return "", err
+			}
+
+			dynamicClient, err := dynamic.NewForConfig(cg)
+			if err != nil {
+				return "", err
+			}
+
+			sb := &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"kind":       "ServiceBinding",
+					"apiVersion": "servicebinding.io/v1beta1",
+					"metadata": map[string]interface{}{
+						"name":      req.NamespacedName.Name,
+						"namespace": req.NamespacedName.Namespace,
+					},
+					"spec": map[string]interface{}{
+						"service": map[string]interface{}{
+							"apiVersion": "v1",
+							"kind":       "Secret",
+							"name":       req.NamespacedName.Name,
+						},
+						"workload": map[string]interface{}{
+							"apiVersion": sclaim.Spec.Application.APIVersion,
+							"kind":       sclaim.Spec.Application.Kind,
+							"selector": map[string]interface{}{
+								"matchLabels": sclaim.Spec.Application.Selector.MatchLabels,
+							},
+						},
+					},
+				}}
+			gvr := schema.GroupVersionResource{
+				Group:    "servicebinding.io",
+				Version:  "v1beta1",
+				Resource: "servicebindings",
+			}
+			_, err = dynamicClient.Resource(gvr).Namespace(req.NamespacedName.Namespace).Create(ctx, sb, metav1.CreateOptions{})
+			if err != nil {
+				return "", err
+			}
+		}
+	}
+	return rsName, nil
+}
+
+func itemContains(s []string, e string) bool {
```

Comment:
I propose to use a dedicated pkg for these kind of functions, like `pkg/slices`.

Reply:
Done